File: bindings/java/src/main/java/dev/invoicekit/RestSidecarEngineClient.java
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 The InvoiceKit Authors

package dev.invoicekit;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Objects;

/** REST sidecar fallback for runtimes where the native library is unavailable. */
public final class RestSidecarEngineClient implements EngineClient {
    /** Header used by the sidecar to preserve the C ABI status code. */
    public static final String STATUS_HEADER = "X-InvoiceKit-Status-Code";

    private final HttpClient httpClient;
    private final URI processEndpoint;

    /**
     * Create a sidecar client for a full Engine ABI process endpoint.
     *
     * @param processEndpoint full sidecar endpoint that accepts Engine ABI JSON bytes.
     */
    public RestSidecarEngineClient(URI processEndpoint) {
        this(HttpClient.newHttpClient(), processEndpoint);
    }

    RestSidecarEngineClient(HttpClient httpClient, URI processEndpoint) {
        this.httpClient = Objects.requireNonNull(httpClient, "httpClient");
        this.processEndpoint = requireAbsolute(processEndpoint);
    }

    @Override
    public int abiVersion() {
        return InvoiceKit.ENGINE_ABI_VERSION;
    }

    @Override
    public EngineResult process(byte[] requestBytes) throws InvoiceKitException {
        Objects.requireNonNull(requestBytes, "requestBytes");
        HttpRequest request = HttpRequest.newBuilder(processEndpoint)
                .timeout(Duration.ofSeconds(30))
                .header("Content-Type", "application/json")
                .header("Accept", "application/json")
                .POST(HttpRequest.BodyPublishers.ofByteArray(requestBytes))
                .build();
        HttpResponse<byte[]> response;
        try {
            response = httpClient.send(request, HttpResponse.BodyHandlers.ofByteArray());
        } catch (IOException error) {
            throw new InvoiceKitException(
                    "sidecar_io_error",
                    "InvoiceKit REST sidecar request failed",
                    "Start the InvoiceKit REST sidecar and pass its Engine ABI endpoint URI.",
                    error);
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
            throw new InvoiceKitException(
                    "sidecar_interrupted",
                    "InvoiceKit REST sidecar request was interrupted",
                    "Retry the request after the calling thread is allowed to run.",
                    error);
        }
        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new InvoiceKitException(
                    "sidecar_http_error",
                    "InvoiceKit REST sidecar returned HTTP " + response.statusCode(),
                    "Check the sidecar URL, health endpoint, and server logs.");
        }
        int statusCode = response.headers()
                .firstValue(STATUS_HEADER)
                .map(RestSidecarEngineClient::parseStatusHeader)
                .orElse(EngineResult.STATUS_OK);
        return new EngineResult(statusCode, response.body());
    }

    private static URI requireAbsolute(URI endpoint) {
        Objects.requireNonNull(endpoint, "processEndpoint");
        if (!endpoint.isAbsolute()) {
            throw new IllegalArgumentException("processEndpoint must be an absolute URI");
        }
        return endpoint;
    }

    private static int parseStatusHeader(String value) {
        try {
            return Integer.parseUnsignedInt(value);
        } catch (NumberFormatException error) {
            return EngineResult.STATUS_ERROR;
        }
    }
}

File: bindings/dotnet/src/InvoiceKit/RestSidecarEngineClient.cs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 The InvoiceKit Authors

using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace InvoiceKit;

/// <summary>
/// REST sidecar fallback for environments where native loading is unavailable.
/// </summary>
public sealed class RestSidecarEngineClient : IEngineClient
{
    /// <summary>
    /// Header used by the sidecar to preserve the C ABI status code.
    /// </summary>
    public const string StatusHeader = "X-InvoiceKit-Status-Code";

    private readonly HttpClient httpClient;
    private readonly Uri processEndpoint;

    /// <summary>
    /// Create a sidecar client for a full Engine ABI process endpoint.
    /// </summary>
    /// <param name="processEndpoint">Full sidecar endpoint that accepts Engine ABI JSON bytes.</param>
    public RestSidecarEngineClient(Uri processEndpoint)
        : this(new HttpClient(), processEndpoint)
    {
    }

    internal RestSidecarEngineClient(HttpClient httpClient, Uri processEndpoint)
    {
        this.httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        this.processEndpoint = RequireAbsolute(processEndpoint);
    }

    /// <inheritdoc />
    public uint AbiVersion => EngineClients.EngineAbiVersion;

    /// <inheritdoc />
    public EngineResult Process(byte[] requestBytes)
    {
        ArgumentNullException.ThrowIfNull(requestBytes);

        using var request = new HttpRequestMessage(HttpMethod.Post, processEndpoint)
        {
            Content = new ByteArrayContent(requestBytes),
        };
        request.Content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        HttpResponseMessage response;
        try
        {
            response = httpClient.Send(request);
        }
        catch (HttpRequestException error)
        {
            throw new InvoiceKitException(
                "sidecar_io_error",
                "InvoiceKit REST sidecar request failed",
                "Start the InvoiceKit REST sidecar and pass its Engine ABI endpoint URI.",
                error);
        }
        catch (TaskCanceledException error)
        {
            throw new InvoiceKitException(
                "sidecar_timeout",
                "InvoiceKit REST sidecar request timed out",
                "Check that the sidecar is reachable and increase the caller timeout if needed.",
                error);
        }

        using (response)
        {
            using var responseStream = response.Content.ReadAsStream();
            using var responseMemory = new MemoryStream();
            responseStream.CopyTo(responseMemory);
            var responseBytes = responseMemory.ToArray();
            if (!response.IsSuccessStatusCode)
            {
                throw new InvoiceKitException(
                    "sidecar_http_error",
                    $"InvoiceKit REST sidecar returned HTTP {(int)response.StatusCode}",
                    "Check the sidecar URL, health endpoint, and server logs.");
            }

            var statusCode = response.Headers.TryGetValues(StatusHeader, out var values)
                ? ParseStatusHeader(values.FirstOrDefault())
                : DeriveStatusCode(responseBytes);
            return new EngineResult(statusCode, responseBytes);
        }
    }

    /// <summary>
    /// Process an Engine ABI JSON request string.
    /// </summary>
    /// <param name="requestJson">Engine ABI JSON request text.</param>
    /// <returns>Copied Engine ABI response bytes plus status code.</returns>
    /// <exception cref="InvoiceKitException">Thrown when the sidecar cannot process the request.</exception>
    public EngineResult Process(string requestJson)
    {
        ArgumentNullException.ThrowIfNull(requestJson);
        return Process(Encoding.UTF8.GetBytes(requestJson));
    }

    /// <summary>
    /// Release the HTTP client owned by this sidecar client.
    /// </summary>
    public void Dispose()
    {
        httpClient.Dispose();
    }

    private static Uri RequireAbsolute(Uri endpoint)
    {
        ArgumentNullException.ThrowIfNull(endpoint);
        if (!endpoint.IsAbsoluteUri)
        {
            throw new ArgumentException("processEndpoint must be an absolute URI.", nameof(endpoint));
        }

        return endpoint;
    }

    private static EngineStatusCode ParseStatusHeader(string? value)
    {
        return uint.TryParse(value, out var parsed)
            ? (EngineStatusCode)parsed
            : EngineStatusCode.Error;
    }

    private static EngineStatusCode DeriveStatusCode(byte[] responseBytes)
    {
        try
        {
            using var document = JsonDocument.Parse(responseBytes);
            if (!document.RootElement.TryGetProperty("status", out var status))
            {
                return EngineStatusCode.Error;
            }

            return status.GetString() == "ok"
                ? EngineStatusCode.Ok
                : EngineStatusCode.Error;
        }
        catch (JsonException)
        {
            return EngineStatusCode.Error;
        }
    }
}
